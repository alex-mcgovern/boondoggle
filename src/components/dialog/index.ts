export { Dialog } from "./dialog.component";
export type { DialogProps } from "./dialog.component";
