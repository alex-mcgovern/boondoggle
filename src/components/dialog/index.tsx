import * as RadixSlot from "@radix-ui/react-slot";
import clsx from "clsx";
import { createRef, forwardRef, useCallback } from "react";

import { useClickOutside } from "../../hooks/use-click-outside";
import { useOpenDialogWithKeyboard } from "../../hooks/use-open-dialog-with-keyboard";
import { useForwardRef } from "../../hooks/use-forward-ref";
import { Box } from "../box";
import { getDialogStyles } from "./styles.css";

import type { ReactNode } from "react";
import type { BoxProps } from "../box";
import type { DialogPlacementEnum } from "./styles.css";

export type DialogProps = BoxProps & {
	/**
	 * Children to render inside the dialog component.
	 */
	children?: ReactNode;

	/**
	 * Optional CSS class name applied to the outer HTML dialog element.
	 */
	className?: string;

	/**
	 * Whether the dialog is "inert" (can't be focused, can't be clicked.)
	 */
	inert?: boolean;

	/**
	 * Controls the dialog's open state, making it a controlled element. Use alongside `onIsOpenChange`.
	 */
	isOpen?: boolean;

	/**
	 * Callback when the dialog's open state is changed. Use alongside `isOpen`.
	 */
	onIsOpenChange?: (isOpen?: boolean) => void;

	/**
	 * Controls the interaction mode for the dialog trigger node.
	 */
	openOn?: "click" | "hover";

	/**
	 * Controls the dialog's placement relative to the trigger node.
	 */
	placement?: DialogPlacementEnum;

	/**
	 * Whether to open the dialog on an enter keypress while the trigger node is focused.
	 */
	preventOpenOnKeydown?: boolean;

	/**
	 * The react node to act as the trigger for the dialog.
	 */
	triggerNode?: ReactNode;

	/**
	 * `BoxProps` to pass to the outer element that wraps the dialog and it's trigger.
	 */
	wrapperProps?: BoxProps;
};

export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
	(
		{
			children,
			className: userClassName,
			isOpen: controlledIsOpen,
			onIsOpenChange,
			openOn = "click",
			placement,
			preventOpenOnKeydown = false,
			triggerNode,
			wrapperProps,
			...rest
		}: DialogProps,
		ref,
	) => {
		const dialogRef = useForwardRef<HTMLDialogElement>(ref);

		const triggerRef = createRef<HTMLElement>();

		const toggleIsOpen = useCallback(() => {
			if (!dialogRef.current) {
				return;
			}

			if (dialogRef.current?.open) {
				dialogRef.current?.close();
			}

			if (!dialogRef.current?.open) {
				dialogRef.current?.show();
			}

			onIsOpenChange?.(dialogRef.current?.open);
		}, [dialogRef, onIsOpenChange]);

		useOpenDialogWithKeyboard<HTMLElement>({
			callback: toggleIsOpen,
			dialogRef,
			preventOpenOnKeydown,
			triggerRef,
		});

		useClickOutside<HTMLDialogElement, HTMLElement>({
			callback: () => {
				return dialogRef.current?.close();
			},
			contentRef: dialogRef,
			triggerRef,
		});

		return (
			<Box position="relative" {...wrapperProps}>
				<Box background="transparent">
					<RadixSlot.Slot
						onClick={toggleIsOpen}
						onMouseEnter={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onMouseLeave={
							openOn === "hover"
								? () => {
										return dialogRef.current?.close();
								  }
								: undefined
						}
						onPointerEnter={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onPointerLeave={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onPointerOver={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						ref={triggerRef}
					>
						{triggerNode}
					</RadixSlot.Slot>
				</Box>
				<Box
					{...rest}
					as="dialog"
					className={clsx(
						userClassName,
						getDialogStyles({ placement }),
					)}
					open={controlledIsOpen}
					ref={dialogRef}
				>
					<Box>{children}</Box>
				</Box>
			</Box>
		);
	},
);
