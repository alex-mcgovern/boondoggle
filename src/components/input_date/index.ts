export { InputDate } from "./input_date.comp";
export type { InputDateProps } from "./input_date.comp";
