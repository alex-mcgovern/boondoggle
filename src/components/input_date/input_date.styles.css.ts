import { style } from "@vanilla-extract/css";

export const inputDate = style({
  selectors: {
    "&:hover": {
      cursor: "pointer",
    },
  },
  textAlign: "left",
});

export const datePickerClosed = style({
  display: "none",
  visibility: "hidden",
});
