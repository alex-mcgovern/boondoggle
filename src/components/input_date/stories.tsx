import { expect } from "@storybook/jest";
import { userEvent, waitFor, within } from "@storybook/testing-library";

import { InputDate as StoryComp } from ".";
import { LOREM } from "../../../mocks/LOREM.mock";

import type { Meta, StoryObj } from "@storybook/react";

const meta = {
    component: StoryComp,
    title: "Components/InputDate",
} satisfies Meta<typeof StoryComp>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        id: LOREM.id(),
        label: LOREM.label(),
        name: LOREM.name(),
        placeholder: LOREM.placeholder(),
    },
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);

        const buttonElement = canvas.getByRole("button");

        expect(buttonElement).not.toBeNull();

        await waitFor(() => {
            userEvent.click(buttonElement);
        });
    },
};
