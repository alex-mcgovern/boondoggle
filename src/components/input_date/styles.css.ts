import { style } from "@vanilla-extract/css";

import { a11yFocus } from "../../styles/common/a11y.css";
import { getSprinkles } from "../../styles/utils/get_sprinkles.css";

export const datePickerDialogStyle = style([
	a11yFocus,
	getSprinkles({
		background: "background",
		border: "border_default",
		borderRadius: "md",
		boxShadow: "md",
		marginTop: "space_2",
		width: "auto",
	}),
	{
		pointerEvents: "all",
	},
]);

export const inputDateStyle = style({
	selectors: {
		"&:hover": {
			cursor: "pointer",
		},
	},
	textAlign: "left",
});
