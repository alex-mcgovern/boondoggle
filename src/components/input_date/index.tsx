/* eslint-disable react-perf/jsx-no-new-array-as-prop */
import clsx from "clsx";
import { forwardRef, useCallback, useState } from "react";

import { DatePicker } from "../date_picker";
import { Dialog } from "../dialog";
import { FieldActionButtonDate } from "../field_action_button_date";
import { Input } from "../input";
import { datePickerDialogStyle, inputDateStyle } from "./styles.css";

import type {
    WithOptionalLabel,
    WithOptionalPlaceholder,
} from "../../common-types";
import type { InputProps } from "../input";
import type { MouseEvent } from "react";

/**
 * Converts a local date to a UTC date.
 */
function convertLocalToUTCDate(date: Date) {
    return new Date(
        Date.UTC(date.getFullYear(), date.getMonth(), date.getDate())
    );
}

export type InputDateProps = Omit<
    InputProps,
    | "isClearable"
    | "isCopyable"
    | "isVisibilityToggleable"
    | "strClear"
    | "placeholder"
    | "strCopy"
    | "strCopied"
    | "strHide"
    | "strShow"
    | "isVisibilityToggleable"
> &
    WithOptionalLabel &
    WithOptionalPlaceholder & {
        /**
         * Controls whether the date picker is open or not.
         */
        isOpen?: boolean;

        /**
         * The locale to use for formatting the date.
         */
        locale?: Intl.LocalesArgument;

        /**
         * Callback to be called when the date changes.
         */
        onChange?:
            | ((date: string) => unknown)
            | ((date: string) => Promise<unknown>);
    };

export const InputDate = forwardRef<HTMLInputElement, InputDateProps>(
    (
        {
            className: userClassName,
            defaultValue,
            isOpen: controlledIsOpen,
            onChange,
            size,
            slotLeft,
            value,
            wrapperProps,
            ...rest
        }: InputDateProps,
        ref
    ) => {
        const [inputValue, setInputValue] = useState<
            string | number | readonly string[]
        >(value || defaultValue || "");

        const [isOpen, setIsOpen] = useState<boolean | undefined>(
            controlledIsOpen
        );

        const onDayClick = useCallback(
            (_: MouseEvent<HTMLElement>, date: Date) => {
                const utcDate = convertLocalToUTCDate(date);
                const isoString = utcDate.toISOString();
                const isoDate = isoString.slice(0, 10);

                setInputValue(isoDate);

                if (onChange) {
                    onChange(isoDate);
                }
                setIsOpen(false);
            },
            [onChange]
        );

        return (
            <Input
                {...(rest as InputProps)}
                className={inputDateStyle}
                defaultValue={defaultValue}
                onChange={(e) => {
                    return onChange?.(e.target.value);
                }}
                ref={ref}
                size={size}
                slotLeft={slotLeft}
                slotRight={
                    <Dialog
                        className={clsx(userClassName, datePickerDialogStyle)}
                        isOpen={isOpen}
                        placement="bottom-end"
                        triggerNode={
                            // eslint-disable-next-line react-perf/jsx-no-jsx-as-prop
                            <FieldActionButtonDate size={size} />
                        }
                    >
                        <DatePicker onDayClick={onDayClick} />
                    </Dialog>
                }
                type="date"
                value={inputValue}
                wrapperProps={wrapperProps}
            />
        );
    }
);
