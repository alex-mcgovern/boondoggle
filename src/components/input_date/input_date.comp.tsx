import { forwardRef, useCallback, useState } from "react";

import { formatDate } from "../../utils/format_date";
import { DatePicker } from "../date_picker/date_picker.comp";
import { Dialog } from "../dialog/dialog.comp";
import { Input } from "../input/input.comp";
import * as styles from "./input_date.styles.css";

import type { InputProps } from "../input/input.comp";
import type { MouseEvent, Ref } from "react";

export type InputDateValue = {
  humanReadable: string;
  raw: string;
};

export type InputDateProps = Omit<InputProps, "value"> & {
  isOpen?: boolean;
  locale?: Intl.LocalesArgument;
  /**
   * On change handler designed to be used with React Hook Form's `register` method.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onChange?: (...event: any[]) => void;
  rawValueTransformer?: (value: string) => string;
};

export const InputDate = forwardRef(
  (
    {
      className: userClassName,
      locale,
      rawValueTransformer,
      defaultValue,
      isOpen: controlledIsOpen,
      wrapperProps,
      onChange,
      ...rest
    }: InputDateProps,
    ref: Ref<HTMLInputElement>
  ) => {
    const [inputValue, setInputValue] = useState<string>("");
    const [isOpen, setIsOpen] = useState<boolean | undefined>(controlledIsOpen);

    /**
     * Callback for when a day is clicked in the date picker.
     */
    const onDayClick = useCallback(
      (_: MouseEvent<HTMLElement>, date: Date) => {
        setInputValue(date.toLocaleDateString(locale));

        if (onChange) {
          onChange(
            rawValueTransformer
              ? rawValueTransformer(date.toISOString())
              : date.toISOString()
          );
        }

        setIsOpen(false);
      },
      [locale, rawValueTransformer, onChange]
    );

    return (
      <Dialog
        className={userClassName}
        isOpen={isOpen}
        onIsOpenChange={setIsOpen}
        ref={ref}
        triggerNode={
          <Input
            {...(rest as InputProps)}
            className={styles.inputDate}
            defaultValue={defaultValue ? formatDate(defaultValue) : undefined}
            readOnly
            value={inputValue}
          />
        }
        wrapperProps={wrapperProps}
      >
        <DatePicker onDayClick={onDayClick} />
      </Dialog>
    );
  }
);
