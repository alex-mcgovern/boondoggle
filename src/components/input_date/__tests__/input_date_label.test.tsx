/** @jest-environment jsdom */
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { LOREM } from "../../../../mocks/LOREM.mock";
import "../../../../test/dialog.mock";
import { InputDate } from "../input_date.comp";

import type { InputDateProps } from "../input_date.comp";

/** ----------------------------------------------------------------------------- */

const PROPS: InputDateProps = {
  id: LOREM.id(),
  label: LOREM.label(),
  name: LOREM.name(),
  placeholder: LOREM.placeholder(),
};

const renderComponent = (props: InputDateProps) => {
  return {
    user: userEvent.setup(),
    ...render(<InputDate {...props} />),
  };
};

/** ----------------------------------------------------------------------------- */

describe("<InputDate />", () => {
  it("should be labelled", async () => {
    const { container, getByLabelText } = await renderComponent(PROPS);

    const label = container.querySelector("label");

    expect(label).not.toBeNull();
    expect(label?.getAttribute("for")).toBe(PROPS.id);
    expect(label?.textContent).toBe(PROPS.label);
    expect(
      getByLabelText(PROPS.label as string, { selector: "input" })
    ).not.toBeNull();
  });

  it("should match snapshot", async () => {
    const { getByRole } = await renderComponent(PROPS);

    expect(getByRole("textbox")).toMatchSnapshot();
  });
});
