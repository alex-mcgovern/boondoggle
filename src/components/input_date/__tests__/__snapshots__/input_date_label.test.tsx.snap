// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputDate /> should match snapshot 1`] = `
<input
  class="styles_getSlotLeftOffsetStyles__18m4u7iz styles__18m4u7iy get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_display_flex_mobile__eiij7bb0 get_sprinkles_justifyContent_center__eiij7b3h styles_getSlotLeftOffsetStyles_numSlots_1__18m4u7i10 styles_getSlotLeftOffsetStyles_size_md__18m4u7i14 styles_getSlotLeftOffsetStyles_compound_3__18m4u7i19 styles_getInputStyles__1i5hpgs6 styles__1i5hpgs5 get_sprinkles_width_100%_mobile__eiij7bn3 a11y_a11yDisabled__r8he5t1 styles_getInputStyles_hasBorder_true__1i5hpgs8 styles_hasBorder_true__1i5hpgs4 styles_getInputStyles_size_md__1i5hpgsa styles_size_md__1i5hpgs1 element_size_elementFontSize_md__1f6zrv4 element_size__1f6zrv1 get_sprinkles_fontSize_body_md__eiij7b21 get_sprinkles_lineHeight_body_md__eiij7b3n element_size_elementPadding_md__1f6zrva element_size__1f6zrv7 get_sprinkles_paddingLeft_space_3_mobile__eiij7bif get_sprinkles_paddingRight_space_3_mobile__eiij7bk0 get_sprinkles_paddingTop_space_2_mobile__eiij7bli get_sprinkles_paddingBottom_space_2_mobile__eiij7bgr styles_inputDateStyle__tedwa22"
  id="lorem-ipsum"
  name="Name ipsum"
  placeholder="Placeholder ipsum"
  readonly=""
  value=""
/>
`;
